media/multipart/common/src/test/java/io/helidon/media/multipart/common/MultiPartDecoderTest.java

```diff
@@ -0,0 +1,571 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.media.multipart.common;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Flow.Publisher;
+import java.util.concurrent.Flow.Subscriber;
+import java.util.concurrent.Flow.Subscription;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import io.helidon.common.http.DataChunk;
+import io.helidon.common.reactive.Multi;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.CoreMatchers.hasItems;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.jupiter.api.Assertions.fail;
+import static io.helidon.media.multipart.common.BodyPartTest.MEDIA_CONTEXT;
+
+/**
+ * Tests {@link MultiPartDecoder}.
+ */
+public class MultiPartDecoderTest {
+
+    private static final Logger LOGGER = Logger.getLogger(MultiPartDecoder.class.getName());
+
+    @BeforeAll
+    public static void before() {
+        LOGGER.setLevel(Level.ALL);
+    }
+
+    @AfterAll
+    public static void after() {
+        LOGGER.setLevel(Level.INFO);
+    }
+
+    @Test
+    public void testOnePartInOneChunk() {
+        String boundary = "boundary";
+        final byte[] chunk1 = ("--" + boundary + "\n"
+                + "Content-Id: part1\n"
+                + "\n"
+                + "body 1\n"
+                + "--" + boundary + "--").getBytes();
+
+        final CountDownLatch latch = new CountDownLatch(2);
+        Consumer<BodyPart> consumer = (part) -> {
+            latch.countDown();
+            assertThat(part.headers().values("Content-Id"),
+                    hasItems("part1"));
+            DataChunkSubscriber subscriber = new DataChunkSubscriber();
+            part.content().subscribe(subscriber);
+            subscriber.content().thenAccept(body -> {
+                latch.countDown();
+                assertThat(body, is(equalTo("body 1")));
+            });
+        };
+        BodyPartSubscriber testSubscriber = new BodyPartSubscriber(
+                SUBSCRIBER_TYPE.INFINITE, consumer);
+        partsPublisher(boundary, chunk1).subscribe(testSubscriber);
```

Comment:
Decoder looks basically as a flatmap, I may be wrong but it should be possible to leave all the queuing in decoder to flatMap operator and simplify it considerably, so if this works:
```java
        Multi.from(chunksPublisher(chunk1))
                .flatMap(chunk -> {
                    var decoder = MultiPartDecoder.create(boundary, MEDIA_CONTEXT.readerContext());
                    Single.just(chunk).subscribe(decoder);
                    return decoder;
                }).subscribe(testSubscriber);
``` 
Something like this should be possible:
```java
        var parser = new MIMEParser(boundary){
            List<ReadableBodyPart> parse(DataChunk chunk){
                ...
            }
        };
        Multi.from(chunksPublisher(chunk1))
                .flatMapIterable(parser::parse)
                }).subscribe(testSubscriber);
```

Reply:
Sorry I got Decoder/Encoder vertigo, I meant Encoder, to look like re-implemented flatmap:
```java
    private static String encodeParts(String boundary, WriteableBodyPart... parts) throws Exception {
        Multi<DataChunk> pub = Multi.just(parts)
                .flatMap(part -> {
                    var enc = MultiPartEncoder.create(boundary, MEDIA_CONTEXT.writerContext());
                    Single.just(part).subscribe(enc);
                    return enc;
                });
        return ContentReaders.readString(pub, StandardCharsets.UTF_8).get(10, TimeUnit.SECONDS);
    }
```